src/core/config/network_diagnostic.h

```diff
@@ -0,0 +1,85 @@
+/*
+ *  Copyright (c) 2023, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes compile-time configurations for the Network Diagnostics.
+ *
+ */
+
+#ifndef CONFIG_NETWORK_DIAGNOSTIC_H_
+#define CONFIG_NETWORK_DIAGNOSTIC_H_
+
+/**
+ * @def OPENTHREAD_CONFIG_NET_DIAG_VENDOR_NAME
+ *
+ * Specifies the default Vendor Name string.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_NET_DIAG_VENDOR_NAME
+#define OPENTHREAD_CONFIG_NET_DIAG_VENDOR_NAME ""
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_NET_DIAG_VENDOR_MODEL
+ *
+ * Specifies the default Vendor Model string.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_NET_DIAG_VENDOR_MODEL
+#define OPENTHREAD_CONFIG_NET_DIAG_VENDOR_MODEL ""
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_NET_DIAG_VENDOR_SW_VERSION
+ *
+ * Specifies the default Vendor SW Version string.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_NET_DIAG_VENDOR_SW_VERSION
+#define OPENTHREAD_CONFIG_NET_DIAG_VENDOR_SW_VERSION "OpenThread"
```

Comment:
Should we default to something like `PACKAGE_VERSION`?

Reply:
Since we limited to 16 chars for this, it can get tricky as `PACKAGE_VERSION` is often longer
e.g. I see `thread-reference-20200818-2394-g8812f7118` used as default value (which is 41 chars).

Reply:
Yeah, I don't think we need to include the full version. But maybe something like `ot-<githash>`?

Reply:
I think the `VENDOR_SW_VERSION` represents the device firmware version (which should include OT stack and platform code and any app layer, etc). 

We can consider adding another TLV (say `THREAD_STACK_VERSION`) for Thread stack (can auto-set it as "ot-<githash>")? I feel this may be more helpful? Thoughts? (I can add it in this PR or later from a new PR on its own?)

Reply:
Having a separate attribute for Thread SW version makes sense to me.

Reply:
In the spec, there's also separate TLVs for vendor SW version and Thread stack version. These are used in JOIN_FIN.req. It would make sense to reuse these definitions and names in the Diagnostic namespace - i.e. not defining new semantics. The existing TLVs are called "Vendor SW Version TLV" and "Vendor Stack Version TLV".

Reply:
I checked the MeshCop "Vendor Stack Version TLV". It defines specific fields, OUI, major, minor, etc (which are practically not used). I prefer defining a "Thread Stack Version TLV" (as a NetDiag TLV) with a string value (16 char max?). This gives us more flexibility.

Will try to add this later (most likely in this PR).

Reply:
Added a new TLV `Thread Stack Version TLV` (value 30) with string value which returns the result from `otGetVersionString()`. I allowed the length up to 128 char (since the version string can be long).

Reply:
(Was discussed in SPEC-1178) Indeed most useful to use a string format; proposal was for 64 chars max length. Build time might be cut off then but all key info will remain in there. Also a vendor can easily shorten the Git tag name length to prevent cut-off.